File: scanApp/src/main/java/scan/ScanAppUtils.java
package scan;

import java.util.List;

import rx.Observable;
import rx.Subscriber;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;
import scan.listener.ScanListener;

/**
 * by y on 2016/10/20.
 */

public class ScanAppUtils {

    public static final int ALL_APP = 1; //全部app
    public static final int SYSTEM_APP = 2; //系统app
    public static final int NO_SYSTEM_APP = 3; //第三方app

    private Subscription subscription = null;
    /**
     * true:系统app更新后也算在第三方app里面
     * false:只记录用户安装的app
     */
    private boolean isNoSystem = false;

    private static class ScanHolder {

        private static final ScanAppUtils appUtils = new ScanAppUtils();
    }

    public static ScanAppUtils newInstance() {
        return ScanHolder.appUtils;
    }

    public ScanAppUtils setNoSystem(boolean noSystem) {
        isNoSystem = noSystem;
        return this;
    }

    public void start(@ScanTypeMode final int type, final ScanListener scanListener) {
        unsubscribe();
        nullSubscription();
        subscription = Observable
                .create(new ScanObservable(type, scanListener, isNoSystem))
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Subscriber<List<AppModel>>() {

                    @Override
                    public void onStart() {
                        super.onStart();
                        if (scanListener != null) {
                            scanListener.onScanStart();
                        }
                    }

                    @Override
                    public void onCompleted() {
                        unsubscribe();
                        nullSubscription();
                    }

                    @Override
                    public void onError(Throwable e) {
                        unsubscribe();
                        nullSubscription();
                        if (scanListener != null) {
                            scanListener.onScanError(e);
                        }
                    }

                    @Override
                    public void onNext(List<AppModel> list) {
                        if (scanListener != null) {
                            scanListener.onScanSuccess(list);
                        }
                    }
                });
    }


    private void unsubscribe() {
        if (subscription != null && !subscription.isUnsubscribed()) {
            subscription.unsubscribe();
        }
    }


    private void nullSubscription() {
        if (subscription != null) {
            subscription = null;
        }
    }
}

File: ScanApp/app/src/main/java/scan/ScanAppUtils.kt
package scan

import rx.Observable
import rx.Subscriber
import rx.Subscription
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import scan.listener.ScanListener

/**
 * by y on 2016/10/20.
 */

class ScanAppUtils {

    companion object {
        const val ALL_APP = 1 //全部app
        const val SYSTEM_APP = 2 //系统app
        const val NO_SYSTEM_APP = 3 //第三方app
        val instance by lazy { ScanAppUtils() }
    }

    private lateinit var subscription: Subscription

    /**
     * true:系统app更新后也算在第三方app里面
     * false:只记录用户安装的app
     */
    private var isNoSystem = false

    fun setNoSystem(noSystem: Boolean): ScanAppUtils {
        isNoSystem = noSystem
        return this
    }

    fun start(@ScanTypeMode type: Int, scanListener: ScanListener) {
        unsubscribe()
        subscription = Observable
                .create(ScanObservable(type, scanListener, isNoSystem))
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : Subscriber<List<AppModel>>() {

                    override fun onStart() {
                        super.onStart()
                        scanListener.onScanStart()
                    }

                    override fun onCompleted() {
                        unsubscribe()
                    }

                    override fun onError(e: Throwable) {
                        unsubscribe()
                        scanListener.onScanError(e)
                    }

                    override fun onNext(list: List<AppModel>) {
                        scanListener.onScanSuccess(list)
                    }
                })
    }


    private fun unsubscribe() {
        if (::subscription.isInitialized && !subscription.isUnsubscribed) {
            subscription.unsubscribe()
        }
    }
}
